examples/constraints.rs

```diff
@@ -0,0 +1,217 @@
+                    selection = selection.saturating_add(1).min(6)
+                }
+                Char('k') | Char('h') | Up | Left => selection = selection.saturating_sub(1),
+                _ => (),
+            }
+        }
+    }
+}
+
+struct Example {
+    constraints: Vec<Constraint>,
+}
+
+impl Example {
+    fn new<C>(constraints: C) -> Self
+    where
+        C: Into<Vec<Constraint>>,
+    {
+        Self {
+            constraints: constraints.into(),
+        }
+    }
+}
+
+impl Widget for Example {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let [title, legend, area] = area.split(&Layout::vertical([Ratio(1, 3); 3]));
+        let heading = Line::from(
+            Itertools::intersperse(
+                self.constraints.iter().enumerate().map(|(i, c)| {
+                    Span::styled(
+                        format!("{:?}", c),
+                        Style::default().fg(Color::Indexed(i as u8 + 1)),
+                    )
+                }),
+                Span::raw(", "),
+            )
+            .collect::<Vec<Span>>(),
+        );
+        Paragraph::new(heading)
+            .block(Block::default().padding(Padding::new(0, 0, 1, 0)))
+            .render(title, buf);
+        let width = format!("{} px", legend.width);
+        let separator = "-".repeat((legend.width as usize - width.len()) / 2);
+        Paragraph::new(format!("{}{}{}", separator, width, separator))
+            .fg(Color::DarkGray)
+            .alignment(Alignment::Center)
+            .render(legend, buf);
+        buf.set_string(
+            legend.x,
+            legend.y,
+            "<",
+            Style::default().fg(Color::DarkGray),
+        );
+        buf.set_string(
+            legend.x + legend.width - 1,
+            legend.y,
+            ">",
+            Style::default().fg(Color::DarkGray),
+        );
+        let blocks = Layout::horizontal(self.constraints).split(area);
+        for (i, block) in blocks.iter().enumerate() {
+            let text = format!("{} px", block.width);
+            Paragraph::new(text)
+                .alignment(Alignment::Center)
+                .block(Block::bordered().style(Style::default().fg(Color::Indexed(i as u8 + 1))))
+                .render(*block, buf);
+        }
+    }
+}
+
+fn ui(f: &mut Frame, tab: usize) {
+    let [tabs, area] = f
+        .size()
+        .split(&Layout::vertical([Fixed(3), Proportional(0)]));
+    f.render_widget(
+        Tabs::new([
+            "Fixed",
+            "Length",
+            "Percentage",
+            "Ratio",
+            "Proportional",
+            "Min",
+            "Max",
+        ])
+        .block(Block::bordered().title("Constraints"))
+        .highlight_style(Style::default().yellow())
+        .select(tab)
+        .padding("  ", "  "),
+        tabs,
+    );
+    let [area, _] = area.split(&Layout::horizontal([Fixed(80), Proportional(0)]));
+
+    match tab {
+        0 => {
```

Comment:
Can this be an enum?
Small methods would help here too.
0 => render_fixed_examples(...),
1 => render_length_examples(...),

Reply:
Done!

Reply:
Thanks for the suggestion. Let me know if it is over engineered / over refactored now :)